robot_skills/src/robot_skills/robot.py

```diff
@@ -32,38 +33,53 @@ class Robot(object):
-
         self.tf_listener = tf_server.TFClient()
 
         # Body parts
-        self.base = base.Base(self.robot_name, self.tf_listener, wait_service=wait_services)
-        self.torso = torso.Torso(self.robot_name,wait_service=wait_services)
-        self.spindle = self.torso
-        self.leftArm = arms.Arm(self.robot_name, "left", self.tf_listener)
-        self.rightArm = arms.Arm(self.robot_name, "right", self.tf_listener)
-        self.arms = OrderedDict(left=self.leftArm, right=self.rightArm)
-
-        self.head = head.Head(self.robot_name)
+        self.parts = dict()
+        self.parts['base'] = base.Base(self.robot_name, self.tf_listener)
+        self.parts['torso'] = torso.Torso(self.robot_name, self.tf_listener)
+        self.parts['leftArm'] = arms.Arm(self.robot_name, self.tf_listener, side="left")
+        self.parts['rightArm'] = arms.Arm(self.robot_name, self.tf_listener, side="right")
+        self.parts['head'] = head.Head(self.robot_name, self.tf_listener)
 
         # Human Robot Interaction
-        self.lights = lights.Lights(self.robot_name)
-        self.speech = speech.Speech(self.robot_name, wait_services, lambda: self.lights.set_color(1,0,0), lambda: self.lights.set_color(0,0,1))
-        self.hmi = Api("/" + self.robot_name + '/hmi')
-        self.ears = ears.Ears(self.robot_name, lambda: self.lights.set_color(0,1,0), lambda: self.lights.set_color(0,0,1))
-        self.ears._hmi = self.hmi # TODO: when ears is gone, remove this line
-        self.ebutton = ebutton.EButton()
+        self.parts['lights'] = lights.Lights(self.robot_name, self.tf_listener)
+        self.parts['speech'] = speech.Speech(self.robot_name, self.tf_listener,
+                                             lambda: self.lights.set_color(1, 0, 0),
+                                             lambda: self.lights.set_color(0, 0, 1))
+        self.parts['hmi'] = api.Api(self.robot_name, self.tf_listener)
+        self.parts['ears'] = ears.Ears(self.robot_name,
+                                       lambda: self.lights.set_color(0, 1, 0),
+                                       lambda: self.lights.set_color(0, 0, 1))
+
+        self.parts['ebutton'] = ebutton.EButton(self.robot_name, self.tf_listener)
 
         # Reasoning/world modeling
-        self.ed = world_model_ed.ED(self.robot_name, self.tf_listener, wait_service=wait_services)
+        self.parts['ed'] = world_model_ed.ED(self.robot_name, self.tf_listener)
 
         # Miscellaneous
         self.pub_target = rospy.Publisher("/target_location", geometry_msgs.msg.Pose2D, queue_size=10)
         self.base_link_frame = "/"+self.robot_name+"/base_link"
 
-        #Grasp offsets
+        # Grasp offsets
         #TODO: Don't hardcode, load from parameter server to make robot independent.
         self.grasp_offset = geometry_msgs.msg.Point(0.5, 0.2, 0.0)
 
+        # Create attributes from dict
+        for k, v in self.parts.iteritems():
+            setattr(self, k, v)
```

Comment:
This is indeed much nicer than override __getattr__ to dynamically look up the parts: this gives tab completion.

But instantiating them the old way and then putting in a dictionary or list also works. 

You could also iterate over all members and then check if they are an instance of RobotPart with isinstance(part, RobotPart):
```python
for part in [part for part in self.__dict__.values() if isinstance(part, RobotPart):
    part.wait_for_connections(1.0)
```